### ISSUE TITLE
bot does not reconnect if connection breaks

### ISSUE THREAD

**Issue by StalkR** [StalkR/goircbot#11]
If connection drops, e.g. with irc server showing the bot is gone or by issuing a kill, the bot doesn't seem to realize.
I'm expecting a log like: Connection error: dial tcp xxxxxxxxxxxxx - reconnecting in 1 minute

**Comment by StalkR** [comment]
I think I found the bug in goirc: https://github.com/fluffle/goirc/blob/master/client/connection.go#L415
When disconnected, it gets stuck in `shutdown()` on `conn.wg.Wait()` which is why it never dispatches the `disconnected` event.

```
I0331 22:43:54.956117   24244 connection.go:413] irc.shutdown(): Disconnected from server.
I0331 22:43:54.956153   24244 connection.go:415] irc.shutdown(): before sock.Close
I0331 22:43:54.956289   24244 connection.go:417] irc.shutdown(): before close(donn.die)
I0331 22:43:54.956316   24244 connection.go:419] irc.shutdown(): before conn.wg.Wait()
E0331 22:43:56.237079   24244 connection.go:378] irc.send(): use of closed network connection
```

**Comment by StalkR** [comment]
well I retried - this time waiting some time for the bot to be on IRC - and it worked, the event got dispatched to goircbot, but then it's stuck at `b.Channels()` https://github.com/StalkR/goircbot/blob/master/bot/bot.go#L74
```
I0331 23:06:44.292608   24388 connection.go:246] irc.Connect(): Connecting to irc.stalkr.net:6697 with SSL.
I0331 23:06:45.316513   24388 connection.go:280] adding wg for ping
I0331 23:06:45.408733   24388 connection.go:378] irc.rateLimit(): Flood! Sleeping for 2.08 secs.
I0331 23:06:47.484061   24388 connection.go:378] irc.rateLimit(): Flood! Sleeping for 2.07 secs.
I0331 23:09:09.183547   24388 connection.go:319] recv(): done; now calling shutdown
I0331 23:09:09.183568   24388 connection.go:424] irc.shutdown(): Disconnected from server.
I0331 23:09:09.183574   24388 connection.go:426] irc.shutdown(): before sock.Close
I0331 23:09:09.183685   24388 connection.go:428] irc.shutdown(): before close(donn.die)
I0331 23:09:09.183695   24388 connection.go:430] irc.shutdown(): before conn.wg.Wait()
I0331 23:09:09.183706   24388 connection.go:358] runLoop: done
I0331 23:09:09.183723   24388 connection.go:339] ping: done
I0331 23:09:09.183731   24388 connection.go:296] send() is done
I0331 23:09:09.183738   24388 connection.go:432] irc.shutdown(): before dispatch
goircbot: received disconnected
```

**Comment by StalkR** [comment]
So we have at least two bugs, I'll call them 1 and 2.

I'm able to reproduce 1 is I connect and quickly disconnect the bot
```
I0331 23:15:29.976490   24561 connection.go:323] recv(): done; now calling shutdown
I0331 23:15:29.976512   24561 connection.go:428] irc.shutdown(): Disconnected from server.
I0331 23:15:29.976518   24561 connection.go:430] irc.shutdown(): before sock.Close
I0331 23:15:29.976778   24561 connection.go:432] irc.shutdown(): before close(donn.die)
I0331 23:15:29.976789   24561 connection.go:434] irc.shutdown(): before conn.wg.Wait()
I0331 23:15:29.976799   24561 connection.go:362] runLoop: done
I0331 23:15:29.976816   24561 connection.go:343] ping: done
E0331 23:15:31.222537   24561 connection.go:393] irc.send(): use of closed network connection
```
so it seems that conn.send() didn't release its lock https://github.com/fluffle/goirc/blob/master/client/connection.go#L291

**Comment by StalkR** [comment]
ok for 1) this is what happens
- we are in a flood condition, `conn.write()` is blocked on `time.After` because of flood: https://github.com/fluffle/goirc/blob/master/client/connection.go#L367
- disconnect happens, caught by `conn.recv` which calls `conn.shutdown()`: https://github.com/fluffle/goirc/blob/master/client/connection.go#L314
- `conn.write` gets out of its `time.After`, tries `conn.io.WriteString` which fails, then tries to call `irc.shutdown()` (https://github.com/fluffle/goirc/blob/master/client/connection.go#L371) which is blocked on its mutex (https://github.com/fluffle/goirc/blob/master/client/connection.go#L406) - therefore cannot return
- but `conn.write` was called by `conn.send` (https://github.com/fluffle/goirc/blob/master/client/connection.go#L292) so it's still blocked on that, and cannot execute its `defer conn.wg.Done()` - where we are stuck

**Comment by StalkR** [comment]
fix out for 1), now let's try to understand 2)

**Comment by StalkR** [comment]
Well I wasn't logging everything, it doesn't block on `b.Channels()` and tries to reconnect:
```
goircbot: received disconnected
goircbot: got b.channels =  [#dev]
goircbot: sent b.quit true
I0331 23:55:38.189425   25338 connection.go:246] irc.Connect(): Connecting to irc.stalkr.net:6697 with SSL.
```

but for some unknown reason it fails:
```
I0331 23:56:14.948487   25338 connection.go:411] irc.shutdown(): Disconnected from server.
goircbot: received disconnected
goircbot: got b.channels =  []
goircbot: sent b.quit true
I0331 23:56:14.949110   25338 connection.go:246] irc.Connect(): Connecting to irc.stalkr.net:6697 with SSL.
```
I'll check why it fails, maybe throttle or something.

It also leads us to bug 3, `b.channels` is now empty.

**Comment by StalkR** [comment]
Theory: we're racing on `conn.sock` with `conn.initialise()` called from `shutdown()` and also by `connect()`.
And indeed, `go run -race` shows at least 3 races...

**Comment by StalkR** [comment]
Actually fix for 1 is incomplete, it triggers in different situations, so much races :(

**Comment by StalkR** [comment]
I'm adding debug logging to trace calls within goirc/client/connection.go it's a mess :(

**Comment by StalkR** [comment]
Ok got it now, multiple things happening so it was hard to see clear. My fix was missing the unlock in the return case. Sent out a new PR https://github.com/fluffle/goirc/pull/58

**Comment by StalkR** [comment]
About channels being lost on double disconnect, I added a check to only save when not empty: https://github.com/StalkR/goircbot/blob/master/bot/bot.go#L75

**Comment by StalkR** [comment]
Then goircbot triggers another race: when we're in `shutdown()` and dispatching, goircbot will immediately call `Connect()` but goirc hasn't finished doing its `shutdown()` which means `initialise()` hasn't run. We can end up in a situation where we call `Connect()`, get a socket then we resume and call `initialise()` which sets it back to nil.
We should add another mutex in goirc to protect that, but it's just as easy to fix it in goircbot by not reconnecting too quickly: https://github.com/StalkR/goircbot/blob/master/bot/bot.go#L121

**Comment by StalkR** [comment]
One last issue with reconnect, it seems to have lost all information:
```
I0401 01:21:47.700990   32292 connection.go:414] irc.shutdown(): Disconnected from server.
I0401 01:21:48.701247   32292 connection.go:248] irc.Connect(): Connecting to irc.stalkr.net:6697 with SSL.
I0401 01:21:49.715255   32292 connection.go:384] -> NICK godev
I0401 01:21:49.715401   32292 connection.go:384] -> USER  12 * :
I0401 01:21:49.715663   32292 connection.go:321] <- :irc.stalkr.net NOTICE AUTH :*** Looking up your hostname...
I0401 01:21:49.715679   32292 connection.go:321] <- :irc.stalkr.net NOTICE AUTH :*** Found your hostname (cached)
I0401 01:21:49.755733   32292 connection.go:321] <- PING :BF0EB246
I0401 01:21:49.755759   32292 connection.go:321] <- :irc.stalkr.net 461 godev USER :Not enough parameters
I0401 01:21:49.755906   32292 connection.go:384] -> PONG :BF0EB246
```

**Comment by StalkR** [comment]
I believe this is because
1) `shutdown()` calls `initialise()` which wipes the state: https://github.com/fluffle/goirc/blob/master/client/connection.go#L214
2) `Me()` always takes it from the state: https://github.com/fluffle/goirc/blob/master/client/connection.go#L174
So in the end we've lost `Me.Nick`, `Me.Ident`, `Me.Name` which are needed for `register`: https://github.com/fluffle/goirc/blob/master/client%2Fhandlers.go#L39

**Comment by StalkR** [comment]
Fix: https://github.com/fluffle/goirc/pull/59

Now with both pull requests and my goircbot changes, the bot is now able to reconnect after disconnection!
